### ISSUE TITLE
Error during scoring due to incorrectly formatted test_mentions.txt file

### ISSUE THREAD

**Issue by cbiehl** [huggingface/neuralcoref#208]
I am trying to re-train the model. I successfully extracted the CoNLL / OntoNotes data following the training instructions and set the `use_gold_mentions` parameter to true in the [conllparser.py](https://github.com/huggingface/neuralcoref/blob/master/neuralcoref/train/conllparser.py#L309) script. I experienced an issue in the function [build_and_gather_multiple_arrays](https://github.com/huggingface/neuralcoref/blob/master/neuralcoref/train/conllparser.py#L632), where for some docs the `arrays_dicts` did not seem to contain a tuple or list of length 3 but of length 0. I skipped these docs with a [continue](https://github.com/cbiehl/neuralcoref/blob/master/neuralcoref/train/conllparser.py#L648) statement in the for loop. After this change I could succesfully run the script and generate the training (.npy) files.

I am using the following command to train the model:
`python learn.py --train /mydir/datasets/conll-2012/v4/data/train/ --eval /mydir/datasets/conll-2012/v4/data/development/`

The file `test_mentions.txt` generated by [build_test_file](https://github.com/huggingface/neuralcoref/blob/master/neuralcoref/train/evaluator.py#L154) contains incorrect mention annotations:
```
somedocid 000 0 For -
somedocid 000 1 small -
somedocid 000 2 as -
somedocid 000 3 the (3
somedocid 000 4 mustard 3)
somedocid 000 5 seed -
somedocid 000 6 is -
somedocid 000 7 , 4)                <-- incorrect mention annotation
somedocid 000 8 it (4
somedocid 000 9 grows -
somedocid 000 10 into -
```

Is the fact that the `arrays_dict` variable contained empty entries (i.e. of length 0) in [build_and_gather_multiple_arrays](https://github.com/huggingface/neuralcoref/blob/master/neuralcoref/train/conllparser.py#L632) related to the incorrect mention annotation? 
Is there any change I could make to the training scripts (e.g. `conllparser.py` or `evaluator.py`) to avoid the incorrect mention annotations?

Thank you in advance for any guidance or hints.
Here is the full log output when using the `learn.py` script as described above:
```
Namespace(all_pairs_epoch=200, all_pairs_l2=1e-06, all_pairs_lr=0.0002, batchsize=20000, checkpoint_file=None, conll_eval_interval=10, conll_train_interval=20, costfl=0.4, costfn=0.4, costs={'FN': 0.4, 'FL': 0.4, 'WL': 1.0}, costwl=1.0, cuda=False, eval='mydir2/datasets/conll-2012/v4/data/development//numpy/', evalkey='mydir2/datasets/conll-2012/v4/data/development//key.txt', h1=1000, h2=500, h3=500, log_interval=10, min_lr=2e-08, numworkers=8, on_eval_decrease='nothing', patience=3, ranking_epoch=200, ranking_l2=1e-05, ranking_lr=2e-06, save_path='mydir/neuralcoref/neuralcoref/train/checkpoints/Oct10_16-01-18_mycomputer.local_', seed=1111, startstage=None, startstep=None, top_pairs_epoch=200, top_pairs_l2=1e-05, top_pairs_lr=0.0002, train='mydir2/datasets/conll-2012/v4/data/train//numpy/', trainkey='mydir2/datasets/conll-2012/v4/data/train//key.txt', weights=None)
Training for 200 200 200 epochs
loading mydir2/datasets/conll-2012/v4/data/train//numpy/tuned_word_embeddings.npy
torch.Size([34288, 50])
loading mydir2/datasets/conll-2012/v4/data/train//numpy/tuned_word_vocabulary.txt
🏝 Loading Dataset at mydir2/datasets/conll-2012/v4/data/train//numpy/
Reading mentions_features.npy, pairs_labels.npy, mentions_labels.npy, pairs_features.npy, mentions_words.npy, static_word_embeddings.npy, pairs_ant_index.npy, mentions_pairs_start_index.npy, mentions_spans.npy, tuned_word_embeddings.npy, mentions_pairs_length.npy, 🏝 Loading Dataset at mydir2/datasets/conll-2012/v4/data/development//numpy/
Reading mentions_features.npy, pairs_labels.npy, mentions_labels.npy, pairs_features.npy, mentions_words.npy, static_word_embeddings.npy, pairs_ant_index.npy, mentions_pairs_start_index.npy, mentions_spans.npy, tuned_word_embeddings.npy, mentions_pairs_length.npy, Vocabulary: 34288
🏝 Build model
mydir/neuralcoref/neuralcoref/train/model.py:34: UserWarning: nn.init.uniform is now deprecated in favor of nn.init.uniform_.
  nn.init.uniform(self.word_embeds.weight.data, a=-0.5, b=0.5)
mydir/neuralcoref/neuralcoref/train/model.py:36: UserWarning: nn.init.xavier_uniform is now deprecated in favor of nn.init.xavier_uniform_.
  nn.init.xavier_uniform(t)
mydir/neuralcoref/neuralcoref/train/model.py:38: UserWarning: nn.init.constant is now deprecated in favor of nn.init.constant_.
  nn.init.constant(t, 0)
🏝 Loading conll evaluator
Preparing batches 📚
Dataset has: 98 batches, 36656 mentions, 1900309 pairs
Reading locations.bin, spacy_lookup.bin, doc.bin, conll_tokens.bin, Done
Preparing batches 📚
Dataset has: 826 batches, 286004 mentions, 16257868 pairs
Reading locations.bin, spacy_lookup.bin, doc.bin, conll_tokens.bin, Done
🏝 Testing evaluator and getting first eval score
🌋 Test evaluator / print all mentions
🌋 Building test file
🌋 Construct test file
Writing in mydir/neuralcoref/neuralcoref/train/test_mentions.txt
🌋 Computing score
Error during the scoring
Command '['perl', 'mydir/neuralcoref/neuralcoref/train/scorer_wrapper.pl', 'muc', 'mydir2/datasets/conll-2012/v4/data/development//key.txt', 'mydir/neuralcoref/neuralcoref/train/test_mentions.txt']' returned non-zero exit status 25.
Detected the end of a mention [1](1) without begin (?,11) at mydir/neuralcoref/neuralcoref/train/scorer/lib/CorScorer.pm line 294, <F> line 1086.
version: 8.01 mydir/neuralcoref/neuralcoref/train/scorer/lib/CorScorer.pm

Traceback (most recent call last):
  File "learn.py", line 325, in <module>
    run_model(args)
  File "learn.py", line 136, in run_model
    eval_evaluator.test_model()
  File "mydir/neuralcoref/neuralcoref/train/evaluator.py", line 152, in test_model
    self.get_score(file_path=ALL_MENTIONS_PATH)
  File "mydir/neuralcoref/neuralcoref/train/evaluator.py", line 226, in get_score
    file_path], stderr=subprocess.STDOUT, encoding='utf-8')
  File "/anaconda3/envs/masterthesis/lib/python3.6/subprocess.py", line 356, in check_output
    **kwargs).stdout
  File "/anaconda3/envs/masterthesis/lib/python3.6/subprocess.py", line 438, in run
    output=stdout, stderr=stderr)
subprocess.CalledProcessError: Command '['perl', 'mydir/neuralcoref/neuralcoref/train/scorer_wrapper.pl', 'muc', 'mydir2/datasets/conll-2012/v4/data/development//key.txt', 'mydir/neuralcoref/neuralcoref/train/test_mentions.txt']' returned non-zero exit status 25.
```

**Comment by svlandeg** [comment]
Hi @cbiehl , I ran into similar issues with `use_gold_mentions=True`. Perhaps this comment is useful: 
Issue #160, but I haven't been able to check this yet. 

For now I'm trying to fix the training with `use_gold_mentions=False`, which requires fixing the non-existing call to `self._extract_mentions` (Work in progress: PR #211 - any feedback welcome)

**Comment by cbiehl** [comment]
Hi @svlandeg, #160 solved my problem, thanks a lot! I will have a look at the PR.

Just for completeness:
I also had to change `loss.data[0]` to `loss.item()` (I am on torch version 1.2.0, should be necessary from torch version 0.5 onwards because of the [merge of Variable and Tensor](https://pytorch.org/docs/stable/autograd.html#variable-deprecated) according to [PyTorch \#6061](https://github.com/pytorch/pytorch/issues/6061).

**Comment by svlandeg** [comment]
@cbiehl : great to hear that solves your problem, and thanks for the additional pointers! I'll work those into a PR to update the code (unless you want to go ahead?)

**Comment by cbiehl** [comment]
I can open a PR, I could also include for some other cases of deprecations with torch>=0.5 if compability with torch==0.4.0 is not a requirement.

**Comment by svlandeg** [comment]
I think that's fine if you set a version in `requirements.txt` accordingly!

**Comment by svlandeg** [comment]
Original issue fixed by PR #213 - thanks @cbiehl and @rupimanoj !

@cbiehl : if you would like to address the Torch changes for the training part, feel free to do in a separate PR!
